File: backend/src/models/Alert.js
const mongoose = require('mongoose');

const alertSchema = new mongoose.Schema(
  {
    alertId: {
      type: String,
      required: true,
      unique: true,
      index: true
    },
    timestamp: {
      type: Date,
      default: Date.now,
      index: true
    },
    sourceIp: {
      type: String,
      default: '0.0.0.0',
      index: true
    },
    destinationIp: {
      type: String,
      default: '0.0.0.0'
    },
    sourcePort: {
      type: Number,
      default: 0
    },
    destinationPort: {
      type: Number,
      default: 0
    },
    protocol: {
      type: String,
      default: 'TCP'
    },
    attackType: {
      type: String,
      required: true,
      index: true
    },
    category: {
      type: String,
      default: 'General Anomaly'
    },
    severity: {
      type: String,
      enum: ['INFO', 'LOW', 'MEDIUM', 'HIGH', 'CRITICAL'],
      default: 'MEDIUM',
      index: true
    },
    confidenceScore: {
      type: Number,
      default: 0.90
    },
    riskScore: {
      type: Number,
      default: 50
    },
    modelUsed: {
      type: String,
      default: 'Random Forest'
    },
    status: {
      type: String,
      enum: ['DETECTED', 'NEW', 'ACKNOWLEDGED', 'INVESTIGATING', 'RESOLVED', 'CLOSED'],
      default: 'NEW',
      index: true
    },
    assignedTo: {
      type: String,
      default: null
    },
    assignedToName: {
      type: String,
      default: null
    },
    incidentId: {
      type: String,
      default: null,
      index: true
    },
    description: {
      type: String,
      default: ''
    },
    recommendation: {
      type: String,
      default: ''
    },
    firstDetectedAt: {
      type: Date,
      default: Date.now
    },
    lastDetectedAt: {
      type: Date,
      default: Date.now
    },
    occurrenceCount: {
      type: Number,
      default: 1
    },
    isDemoData: {
      type: Boolean,
      default: false
    }
  },
  {
    timestamps: true
  }
);

// Compound indexes for query optimization
alertSchema.index({ timestamp: -1, severity: 1 });
alertSchema.index({ sourceIp: 1, attackType: 1, status: 1 });

module.exports = mongoose.model('Alert', alertSchema);

File: backend/app/models/alert.py
import uuid
import enum
from typing import Optional, List, TYPE_CHECKING
from sqlalchemy import String, Integer, Enum, ForeignKey, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.db.base import Base

if TYPE_CHECKING:
    from app.models.traffic import TrafficFlow
    from app.models.user import User
    from app.models.incident import Incident

class AlertSeverity(str, enum.Enum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"
    CRITICAL = "CRITICAL"

class AlertStatus(str, enum.Enum):
    NEW = "NEW"
    ACKNOWLEDGED = "ACKNOWLEDGED"
    INVESTIGATING = "INVESTIGATING"
    RESOLVED = "RESOLVED"
    FALSE_POSITIVE = "FALSE_POSITIVE"

class Alert(Base):
    __tablename__ = "alerts"

    alert_id: Mapped[str] = mapped_column(String(50), unique=True, nullable=False, index=True)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    alert_type: Mapped[str] = mapped_column(String(100), nullable=False, index=True)
    severity: Mapped[AlertSeverity] = mapped_column(Enum(AlertSeverity, name="alert_severity_enum"), nullable=False, index=True)
    status: Mapped[AlertStatus] = mapped_column(Enum(AlertStatus, name="alert_status_enum"), default=AlertStatus.NEW, nullable=False, index=True)
    risk_score: Mapped[int] = mapped_column(Integer, nullable=False, index=True)

    flow_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("traffic_flows.id", ondelete="CASCADE"), nullable=False, index=True)
    flow: Mapped["TrafficFlow"] = relationship("TrafficFlow", back_populates="alerts")

    assigned_to_id: Mapped[Optional[uuid.UUID]] = mapped_column(ForeignKey("users.id"), nullable=True)
    assigned_to: Mapped[Optional["User"]] = relationship("User")

    incident_id: Mapped[Optional[uuid.UUID]] = mapped_column(ForeignKey("incidents.id", ondelete="SET NULL"), nullable=True)
    incident: Mapped[Optional["Incident"]] = relationship("Incident", back_populates="alerts")

    notes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
